multicluster/controllers/multicluster/stale_controller.go

```diff
@@ -86,6 +92,11 @@ func NewStaleResCleanupController(
 // +kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=resourceexports,verbs=get;list;watch;delete
 
 func (c *StaleResCleanupController) cleanup() error {
+	if mutexassert.RWMutexRLocked(c.mutex) {
```

Comment:
Why we need to back off, rather than blocking on the mutex?

Reply:
We don't have to. However, Lan and I believe that since stale controller performs a lower priority operation, it is better to give way to more time critical operations that ResourceImport reconcilers are performing. For example, on mc controller restart, there might be hundreds of LabelIdentity ResourceImports to be synced. We do not want that to be blocked by the cleanup function, which makes quite a number of apiserver rest calls.

Reply:
I am not sure if that helps much. But we can go with this approach for now.

Reply:
I also wonder if it's really necessary to introduce a mutex to all these controllers and a new 3rd party lib for inspecting the lock's state. I think we encountered the situation the patch tries to fix a few times. A common solution is, taking the snapshots of the resources being garbage collected in right order. For example, if type A triggers type B's creation, then the garbage collector should take snapshot of type B first, then type A. It can guarantee non of B resources will be marked as stale incorrectly.
For example:
At t1, A1 is created;
At t2, B1 is created;
At t3, A2 is created;
At t4, B2 is created.

As long as we collect B first, non of B resouces would be marked stale regardless of which time the collection is executed.
It could happen that stale controller and the resource's own controller trying to delete one resource at the same time when A is removed, but it's very easy to resolve, by just tolerating NotFound error on deletion call.

Reply:
Thanks for the suggestion Quan!! I've re-written the patch the way you suggested and verified it works -- the testcase added in this PR will not pass with ToT implementation, but passes with either the mutex solution or with the order of snapshot flipped for garage collecting. I agree that the later is a cleaner and better approach, so updated the PR accordingly. Also, the `NotFound` error in reconciling deleted ResourceImports is already ignored in the current implementation.